### ISSUE TITLE
Update to use grahpql-upload v8

### ISSUE THREAD

**Issue by jkosonen** [apollographql/apollo-server#1854]
> You can see in the graphql-upload (formerly apollo-upload-server) changelog that a lot of critical improvements have been made between v5 and v7. Some of the problems were very challenging to solve! It should be stable now and we have a lot of tests.
> 
> Everyone should migrate from apollo-upload-server v5 to graphql-upload v8+ ASAP.
> 

apollo-upload-server should be updated to use latest graphql-upload

**Comment by jaydenseric** [comment]
This is a pretty critical issue. Not only are people missing out on a huge amount of fixes and a much more powerful API, but the public can crash Apollo Server v2 infrastructure by sending malformed multipart requests (in theory; I have not gone around trying it). [People are even talking about filing CVEs](https://github.com/apollographql/apollo-server/issues/1509#issuecomment-432600001)!

Here is a PR that would resolve this issue: https://github.com/apollographql/apollo-server/pull/1764.

I ❤️you all at Apollo, event if this particular issue needs work. I don't chase dramas, I'm just trying to make the world a better place and offer some transparency to the community. It's their apps that are affected, after all 🙏

I'll be in San Francisco next week for [GraphQL Summit](https://summit.graphql.com), and am happy to meet anyone in person that would like to talk about it in person or collaborate.

## Context

Before [`apollo-upload-server`](https://npm.im/apollo-upload-server) (since renamed [`graphql-upload`](https://npm.im/graphql-upload)) was implemented in Apollo Server v2, unsolved challenges and their security implications were transparent to the Apollo and the public via Github issues and Apollo Slack team threads. I also messaged core Apollo team members via Slack about the challenges and gave my best estimates as to when they might be resolved. The response was that as long as they get fixed sometime soonish, not to worry.

It turns out that running resolvers before the request has finished, so that file upload streams can be handled efficiently in resolvers, is super challenging for a lot of reasons. I worked tirelessly on [tests for every edge case](https://github.com/jaydenseric/graphql-upload/blob/v8.0.0/src/test.mjs) and solid solutions to each challenge (requests aborted half way through, the same file being used as an argument for multiple mutations, files upload streams being consumed out of order, etc. etc.) with heroic contributions from @mike-marcacci who created [`fs-capacitor`](https://npm.im/fs-capacitor) as a big part of the solution. All the issues have been resolved! Today there are 94 closed issues, and only 2 open… A question, and a request for Node.js v6 support:

<img width="479" alt="screen shot 2018-10-29 at 10 46 25 pm" src="https://user-images.githubusercontent.com/1754873/47647930-a140f400-dbcc-11e8-8b1d-136bedbf9d58.png">

In the process of solving all the complicated challenges across multiple versions of Node.js (with their subtle and not-so-subtle breaking API changes to streams, etc.), Node.js support was lifted from v6 to v8.5. See https://github.com/jaydenseric/graphql-upload/issues/109. I never thought this was a big deal, since absolutely no one I know in 2018 is using Node.js v6, which is [only a few months away from end-of-life](https://github.com/nodejs/Release#release-schedule), when v8 LTS and v10 LTS are available and vastly superior. It turns out that Apollo _really_ doesn't want to drop Node.js v6.

[I've said that I'm willing to accept a PR adding Node.js v6 support](https://github.com/jaydenseric/graphql-upload/issues/109#issuecomment-413864786), even though it means compromising some modern features that I've come to enjoy, such as native async/await. I've even cleared the path by [dropping support back to Node.js v6 for a dev dependency](https://github.com/jaydenseric/graphql-upload/issues/109#issuecomment-416489577).

Not everyone uses Apollo Server. [`express-graphql`](https://npm.im/express-graphql) gets around 112k downloads per month while [`apollo-server-core`](https://npm.im/apollo-server-core) gets around 150k. Personally I've used most of the GraphQL servers available on npm including Apollo Server, and currently use [`graphql-api-koa`](https://npm.im/graphql-api-koa). [`graphql-upload`](https://npm.im/graphql-upload) supports supports most GraphQL servers. My point is, if something like Node.js v6 support is specifically important to Apollo, a for-profit company, it is the perfect opportunity to make a contribution back to the project via a PR. After all, I have contributed to Apollo repos before and have sent _hundreds_ of messages of free support for the community on Slack; I get notifications anytime the word `upload` is mentioned.

I've spent months of my life dedicated to this problem space at great personal cost, with [no contributions from Apollo](https://github.com/jaydenseric/graphql-upload/graphs/contributors), despite it powering a major feature of Apollo Server, one of their flagship products. All along, I have been motivated by the promise that the work will help not just myself, but a lot of other people handle uploads intuitively and efficiently in their apps. It breaks my heart every day to see a horribly broken fork ([`@apollographql/apollo-upload-server`](https://npm.im/@apollographql/apollo-upload-server)) of the last version that happened to support Node.js v6 (that predates fixes) being forced on a trusting community to the tune of 136k installs per week:

<img width="370" alt="screen shot 2018-10-29 at 10 58 16 pm" src="https://user-images.githubusercontent.com/1754873/47648495-4f996900-dbce-11e8-8ec1-7beee4fdb86b.png">

Users keep coming to me for support about issues we already worked hard to solve _months_ ago. This poor developer experience needlessly tarnishes both my own and Apollo's reputation and really hits morale.

Due to a lack of progress I've tried my best to reach out in the `#apollo-maintainers` Slack channel multiple times…

<img width="615" alt="screen shot 2018-10-30 at 12 01 42 am" src="https://user-images.githubusercontent.com/1754873/47651674-f46c7400-dbd7-11e8-88d8-20773754b6c6.png">

The problem and solution gets acknowledged…

<img width="719" alt="screen shot 2018-10-30 at 12 03 18 am" src="https://user-images.githubusercontent.com/1754873/47651672-f3d3dd80-dbd7-11e8-8bac-11f957fadfdd.png">

Then ignored again…

<img width="917" alt="screen shot 2018-10-29 at 10 18 55 pm" src="https://user-images.githubusercontent.com/1754873/47646747-fed34180-dbc8-11e8-80ba-82a38678c204.png">

## What Apollo Server users can do in the meantime

I have updated the [apollo-upload-examples API](https://github.com/jaydenseric/apollo-upload-examples/tree/master/api) to disable the outdated and dangerous Apollo Server upload  handling and use `graphql-upload` instead; [here is the commit](https://github.com/jaydenseric/apollo-upload-examples/commit/921aa517baeeb5a967bf02788a8ac49d9407d998).

To migrate your existing Apollo Server v2 project:

1. `npm install graphql-upload`.
2. [Disable the Apollo Server upload handling by using `uploads: false` in the `ApolloServer` options](https://github.com/jaydenseric/apollo-upload-examples/blob/master/api/server.mjs#L19).
3. [Setup the `graphql-upload` middleware](https://github.com/jaydenseric/apollo-upload-examples/blob/1adea5abf6348020aa70299f09a42eaf91e9278f/api/server.mjs#L8).
4. Setup the `Upload` scalar resolver.
    - If you're using string typedefs:
      1. [Add `scalar Upload` to the typedefs string](https://github.com/jaydenseric/apollo-upload-examples/blob/1adea5abf6348020aa70299f09a42eaf91e9278f/api/types.mjs#L2).
      2. [Add the `GraphQLUpload` resolver for the `Upload` scalar](https://github.com/jaydenseric/apollo-upload-examples/blob/1adea5abf6348020aa70299f09a42eaf91e9278f/api/resolvers.mjs#L50).
    - If you're using JS objects to define your schema (how I roll), use the [`GraphQLUpload` named export of `graphql-upload`](https://github.com/jaydenseric/graphql-upload#class-graphqlupload) for the `Upload` scalar.
5. Migrate to the new API in your mutation resolvers:
    
    ```diff
    - const { stream, filename, mimetype } = await upload
    + const { createReadStream, filename, mimetype } = await upload
    + const stream = createReadStream()
    ```
    
    If you don't, it will still work for now but you will see this deprecation warning:
    
    <img width="509" alt="screen shot 2018-10-29 at 10 03 43 pm" src="https://user-images.githubusercontent.com/1754873/47646003-b0bd3e80-dbc6-11e8-8ce3-7dd3299a1ed6.png">
    
   This API change [happened in v7](https://github.com/jaydenseric/graphql-upload/blob/master/changelog.md#700) and was nessesary to fix a whole bunch of issues, the most obvious being that you can now use a single `Upload` as a variable that can be used as input for multiple mutations in one request. Previously, consuming the `stream` in one resolver would interfere with consuming the same `stream` in another.

**Comment by MANTENN** [comment]
File returns a promise and inside an array. This is a solution for those who encounter the same issue as I do
   ```
 uploadMaterialImage: async (p, a, { models, user }) => {
      const { env: { PUBLICDIR } } = process;
      const uploadDir = "." + PUBLICDIR + "/images";
        let { file: [file] } = await a;
        return file.then(
          async ({
            filename, mimetype, encoding, createReadStream
          }) => {
            const stream = createReadStream();
            const { ok, errors, payload } =  await (async () => {
                const id = shortid.generate();
                let filepath = `${uploadDir}/${id}-${filename}`;
                return new Promise((resolve, reject) => {
                  stream.on('error', err => {
                    if (stream.truncated) {
                      // Delete the truncated file
                      fs.unlinkSync(filepath);
                    }
                    reject({ok: false, errors: err});
                  });
      
                  stream
                    .pipe(fs.createWriteStream(filepath))
                    .on('error', err => reject({ok: false, errors: err}))
                    .on('finish', () => {
                      filepath = filepath.split(".")
                      filepath.shift();
                      filepath = filepath.join(".")
                      return resolve({ok: true, payload: {
                      path: filepath
                     }})});
                });
              })();
              return {
                ok,
                errors,
                payload
              }
          }
        )
    }
```

**Comment by abernix** [comment]
This should be addressed by #2054.  As I've requested in that PR, I'd really appreciate anyone who is utilizing file uploads to try upgrading to the published `alpha` which updates `graphql-upload` to v8.  I've detailed the progress on this matter extensively in #2054, but the high-bit is that this should be ready to try out now in `apollo-server@2.3.0-alpha.0`.  Please detail any problems (or successes!) you encounter with the upgrade, as the feedback will guide its final release.

Ref: https://github.com/apollographql/apollo-server/pull/2054#issuecomment-444471202

**Comment by abernix** [comment]
The `alpha` releases didn't identify any problems so I've graduated this to the official `apollo-server-*@2.3.0` releases.
